fix(platform-brower): fire Animations events & set animation properties when using async animations

JeanMeche opened the pull request:
2 closely related commits : 

* Animations events registered against the default renderer were'nt registered against the animation renderer once it was loaded. 
* Animations properties set on the default renderer weren't set on the animation renderer once it was loaded.

fixes #52076

JeanMeche committed:
fix(platform-browser): Fire Animations events when using async animations.

Animations events registered against the default renderer weren't registered against the animation renderer once it was loaded. This commit fixes this.

fixes #52076

JoostK commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -104,13 +110,26 @@ export class AsyncAnimationRendererFactory implements RendererFactory2 {
+  private isUsingAnimationRenderer = false;
+
+  // List of event registrered by the default renderer
+  private listenedEvents:
```
If you set this field to `null` you can drop `isUsingAnimationRenderer` altogether.

JeanMeche committed:
fix(platform-browser): set animation properties when using async animations.

Animations properties set on the default renderer weren't set on the animation renderer once it was loaded. This commit fixes this.

AndrewKushnir commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -191,6 +217,11 @@ export class DynamicDelegationRenderer implements Renderer2 {
   listen(target: any, eventName: string, callback: (event: any) => boolean | void): () => void {
+    // We need to keep track of animation events registred by the default renderer
+    // So we can also register them against the animation renderer
+    if (this.animationEvents !== null && eventName.startsWith('@')) {
+      this.animationEvents.push({target, eventName, callback});
+    }
```
Just as a refactoring idea: we can probably have a single array where we collect functions that should be invoked later once "real" renderer is loaded, for example:

```suggestion
    if (this.shouldReplay(eventName)) {
      this.replay.push((renderer: Renderer2) => renderer.listen(target, eventName, callback));
    }
```
The `shouldReplay` function may look like this:
```
private shouldReplay(propOrEventName: string): boolean {
  return this.replay !== null && propOrEventName.startsWith(ANIMATION_PREFIX);
}
```
After that in the `use` function we can do something like this:
```
  use(impl: BaseAnimationRenderer) {
    this.delegate = impl;

    if (this.replay !== null) {
      // Replay queued actions using loaded renderer instance to apply
      // all events and properties collected while loading was in progress.
      for (const fn of this.replay) {
        fn(impl);
      }
      // Set to `null` to indicate that the queue was processed
      // and we no longer need to collect events and properties.
      this.replay = null;
    }
  }
```

AndrewKushnir replied:
A couple additional comments (applicable to the current implementation too):
- we should cleanup the queue in the `destroy` method
- we'd need to account for a situation when loading of the real renderer fails, in which case we should stop collecting callbacks and clear the queue, so that we don't have memory leaks (+ we'd need to output `console.error` and mention that loading failed and no animations would be displayed)

JeanMeche replied:
Great suggestions, 👍

JeanMeche commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -6,15 +6,19 @@
     Pick<typeof import('@angular/animations/browser'), 'ɵcreateEngine'|'ɵAnimationRendererFactory'>;
 
+
+const ANIMATION_PREFIX = '@';
```
I propose that we keep it as is for this PR, but we'll switch to an imported value in an upcomming PR. 

Multiple files out of the scope of this PR would benefit from a single reference for the animation prefix.

AndrewKushnir replied:
Having a copy of the const in this package looks good. My main point was that we'd need to extract the `'@'` into a const (vs having it inlined in the code).

AndrewKushnir reviewed (commented):
@JeanMeche the solution makes sense to me, the code looks good. Could you please resolve the CI failure, so that I can do a final code review pass?

JeanMeche commented:
@AndrewKushnir All green for you final review, I just rebased.

jessicajaniuk reviewed (approved):
LGTM! This looks great!

reviewed-for: fw-animations, fw-core, public-api

AndrewKushnir commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -35,8 +38,11 @@ export class AsyncAnimationRendererFactory implements RendererFactory2 {
-          throw new Error('Failed to load the @angular/animations/browser module');
+          throw new RuntimeError(
+              RuntimeErrorCode.ANIMATION_RENDERER_ASYNC_LOADING_FAILURE,
+              'You are using the `provideAnimationsAsync()` provider. ' +
+                  'Angular failed to load the animation renderer. ' +
+                  'No animations will be displayed and their style won\'t be applied.');
```
It make sense to keep the error in prod bundles, but we can tree-shake-away the error message:
```suggestion
              (typeof ngDevMode === 'undefined' || ngDevMode) && 'Async loading for animations package was ' + 
                  'enabled, but loading failed. Angular falls back to using regular rendering. ' +
                  'No animations will be displayed and their styles won\'t be applied.');
```
(I've also rephrased the error message a bit)

If we can safely fall back to a regular rendering, I'm wondering if this `throw new RuntimeError` should be a `console.error` instead? WDYT? We should not change it in this PR though, since this is unrelated to the change.

JeanMeche replied:
Yes, I'll convert it to a standard `console.error` in an upcomming PR.

AndrewKushnir commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -78,10 +84,15 @@ export class AsyncAnimationRendererFactory implements RendererFactory2 {
+          dynamicRenderer.use(animationRenderer);
+        })
+        .catch(e => {
+          dynamicRenderer.handleRendererLoadingError();
```
The delegation renderer should *not* be aware of the loading states (since it's unrelated to that abstraction). I think a better option would be to call `use` with an instance of the current renderer. In this case we permanently start using that fallback renderer. From the delegation renderer perspective, this operation is no different that applying animations renderer.
```suggestion
          // Permanently use regular renderer when loading fails.
          dynamicRenderer.use(<fallback_renderer_reference>);
```

AndrewKushnir reviewed (approved):
@JeanMeche great work 👍 Thanks for addressing the feedback!

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented on packages/platform-browser/src/platform-browser.ts:
```diff
@@ -79,6 +79,7 @@ export {EVENT_MANAGER_PLUGINS, EventManager, EventManagerPlugin} from './dom/eve
 export {HAMMER_GESTURE_CONFIG, HAMMER_LOADER, HammerGestureConfig, HammerLoader, HammerModule} from './dom/events/hammer_gestures';
 export {DomSanitizer, SafeHtml, SafeResourceUrl, SafeScript, SafeStyle, SafeUrl, SafeValue} from './security/dom_sanitization_service';
 export {HydrationFeature, provideClientHydration, HydrationFeatureKind, withHttpTransferCacheOptions, withNoHttpTransferCache} from './hydration';
+export {RuntimeErrorCode as ɵRuntimeErrorCode} from './errors';
```
This export should be in the `packages/platform-browser/src/private_export.ts` file.

AndrewKushnir commented:
[Presubmit](http://test/OCL:572095082:BASE:572095099:1696901063229:b0435181).

AndrewKushnir commented:
**Caretaker note**: presubmit is "green", the saucelabs CI failure is unrelated (connection issue). This PR is ready for merge.
